packages/web-components/src/drawer/drawer.ts

```diff
@@ -0,0 +1,184 @@
+ *
+ * @csspart dialog - The dialog element that represents the drawer.
+ *
+ * @slot - Default slot for the drawer's content.
+ *
+ * @fires dismiss - Emitted when the drawer is dismissed.
+ *
+ * @method connectedCallback() - Called when the custom element is connected to the document's DOM.
+ * @method disconnectedCallback() - Called when the custom element is disconnected from the document's DOM.
+ * @method show() - Opens the drawer if it is currently hidden.
+ * @method hide() - Closes the drawer if it is currently open.
+ * @method clickHandler(event) - Handles click events on the drawer.
+ * @method keydownHandler(event) - Handles keydown events on the drawer.
+ * @method openChanged(oldValue, newValue) - Handles changes to the `open` property.
+ * @method typeChanged(oldValue, newValue) - Handles changes to the `type` attribute.
+ *
+ * @summary A flexible drawer component that can be used in various configurations such as modal, non-modal, alert, inline, overlay, with different sizes and positions.
+ *
+ * @tag fluent-drawer
+ */
+
+export class Drawer extends FASTElement {
+  /**
+   * The dialog element.
+   * @public
+   *
+   */
+  @observable
+  public dialog!: HTMLDialogElement;
+
+  /**
+   * Determines whether the drawer should be displayed as modal, non-modal, or alert.
+   * When in modal or alert mode, an overlay is applied over the rest of the view.
+   * @public
+   */
+  @attr({ attribute: 'type' })
+  public type: DrawerModalType = DrawerModalType.modal;
+
+  /**
+   * The ID of the element that labels the drawer.
+   * @public
+   */
+  @attr({ attribute: 'aria-labelledby' })
+  public ariaLabelledby?: string;
+
+  /**
+   * The ID of the element that describes the drawer.
+   * @public
+   */
+  @attr({ attribute: 'aria-describedby' })
+  public ariaDescribedby?: string;
+
+  /**
+   * Sets the drawer as inline
+   * @public
+   * @defaultValue false
+   */
+  @attr({ mode: 'boolean' })
+  public inline?: boolean = false;
+
+  /**""
+   * Sets the position of the drawer (start/end).
+   * @public
+   * @defaultValue start
+   */
+  @attr
+  public position: DrawerPosition = DrawerPosition.start;
+
+  /**
+   * Sets the size of the drawer (small/medium/large).
+   * @public
+   * @defaultValue medium
+   */
+  @attr({ attribute: 'size' })
+  public size: DrawerSize = DrawerSize.medium;
+
+  /**
+   * Method to emit an event when the dialog is dismissed
+   */
+  public dismiss(): void {
+    this.$emit('dismiss');
+  }
+
+  /**
+   * Method to emit an event when the dialog's open state changes
+   * @public
+   */
+  public emitOpenChange = (): void => {
+    this.$emit('open', { open: this.dialog.open });
+  };
+
+  /**
+   * @public
+   * Method called when the 'type' attribute changes
+   */
+  public inlineChanged(oldValue: boolean, newValue: boolean): void {
+    if (this.$fastController.isConnected) {
+      if (newValue) {
+        this.type = DrawerModalType.nonModal;
+      }
+    }
+  }
```

Comment:
There is an interesting question here on how we should manage this - should we _change_ attribute values from within the component or should we guard against things not ending up broken within the code itself. IE, what takes precedence here? Is it inline? Should we guard against specific roles and behaviors if inline is set and ensure that `inline` is always correct? OR should we guard against inline working if the roles and other configurations aren't properly setup?

Reply:
Thanks for bringing this up. I wasn't thrilled about making changes within the component that the user might not be aware of.

I thinkin guarding against specific roles and behaviors probably has more pros..

The behavior is more predictable to consumers
Better alignment with separating concerns ( validating the state rather than silently adjust attributes.
Error handling informs the developer of the issue
Better state consistency by preventing invalid configs in the first place

I think some issues with the current implementation are..

potential unwanted hidden side effects
Automatically adjusting attributes can lead to inconsistent states, especially if there are multiple interdependent attributes
Lack of developer control

Do we thinking throwing a console error is enough here to guard against this?

Something like:

```ts
  private validateConfiguration(): void {
    if (this.inline && this.type !== DrawerModalType.nonModal) {
      throw new Error('Invalid configuration: inline requires the type to be nonModal');
    }
  }
```